syntax = "proto3";

message RegisterSellerRequest {
  // Seller's address for registration
  string seller_address = 1;
  // Seller's UUID for registration
  string seller_uuid = 2;
}

message RegisterSellerResponse {
  // Result of the registration operation (SUCCESS or FAIL)
  string result = 1;
}

message SellItemRequest {
  // Product details for posting a new item
  string product_name = 1;
  string category = 2;
  int32 quantity = 3;
  string description = 4;
  string seller_address = 5;
  string seller_uuid = 6;
  double price_per_unit = 7;
}

// message SellItemRequest {
//   // Product details for posting a new item
//   string product_name = 1;

//   // Enum for category
//   enum ItemCategory {
//     ELECTRONICS = 0;
//     FASHION = 1;
//     OTHERS = 2;
//   }

//   oneof category_option {
//     ItemCategory category = 2;
//   }

//   int32 quantity = 3;
//   string description = 4;
//   string seller_address = 5;
//   string seller_uuid = 6;
//   double price_per_unit = 7;
// }

message SellItemResponse {
  // Result of the SellItem operation (SUCCESS or FAIL)
  string result = 1;
  // Unique item ID assigned on successful posting (optional)
  int32 item_id = 2;
}

// Message structure for UpdateItem request
message UpdateItemRequest {
  // ID of the item to update
  int32 item_id = 1;
  // New price for the item
  double new_price = 2;
  // New quantity for the item
  int32 new_quantity = 3;
  // Seller's address for credential verification
  string seller_address = 4;
  // Seller's UUID for credential verification
  string seller_uuid = 5;
}

// Message structure for UpdateItem response
message UpdateItemResponse {
  // Result of the update operation (SUCCESS or FAIL)
  string result = 1;
}

// Message structure for DeleteItem request
message DeleteItemRequest {
  // ID of the item to delete
  int32 item_id = 1;
  // Seller's address for credential verification
  string seller_address = 2;
  // Seller's UUID for credential verification
  string seller_uuid = 3;
}

// Message structure for DeleteItem response
message DeleteItemResponse {
  // Result of the delete operation (SUCCESS or FAIL)
  string result = 1;
}

// Message structure for DisplaySellerItems request
message DisplaySellerItemsRequest {
  // Seller's address for displaying items
  string seller_address = 1;
  // Seller's UUID for credential verification (optional)
  string seller_uuid = 2;
}

// Message structure for DisplaySellerItems response
message DisplaySellerItemsResponse {
  // List of items uploaded by the seller
  repeated Item items = 1;

  // Define the structure for the Item message
  message Item {
    // Item ID
    int32 id = 1;
    // Item price
    double price = 2;
    // Item name
    string name = 3;
    // Item category
    string category = 4;
    // Item description
    string description = 5;
    // Quantity remaining
    int32 quantity_remaining = 6;
    // Seller's address
    string seller = 7;
    // Item rating
    double rating = 8;
  }
}

message ItemNotificationRequest2 {
  int32 id = 1;
  float price = 2;
  string name = 3;
  string category = 4;
  string description = 5;
  int32 quantity_remaining = 6;
  string seller = 7;
  float rating = 8;
}


message ItemNotificationResponse2 {
  string result = 1;
}

message GetItemDetailsRequest {
  // Define any request parameters, such as item_id
  int32 item_id = 1;
}

message GetItemDetailsResponse {
  // Define the response fields based on the item details
  int32 id = 1;
  float price = 2;
  string name = 3;
  string category = 4;
  string description = 5;
  int32 quantity_remaining = 6;
  string seller = 7;
  float rating = 8;
  // Add any other fields as needed
}

// gRPC service definition with a different name
service SellerService {
  // ... (other RPC methods)

  // RPC method for registering a seller
  rpc RegisterSeller(RegisterSellerRequest) returns (RegisterSellerResponse);

  // RPC method for selling an item
  rpc SellItem(SellItemRequest) returns (SellItemResponse);

  // RPC method for updating an item
  rpc UpdateItem(UpdateItemRequest) returns (UpdateItemResponse);

  // RPC method for deleting an item
  rpc DeleteItem(DeleteItemRequest) returns (DeleteItemResponse);

  // RPC method for displaying seller items
  rpc DisplaySellerItems(DisplaySellerItemsRequest) returns (DisplaySellerItemsResponse);

  rpc NotifyItemUpdate2(ItemNotificationRequest2) returns (ItemNotificationResponse2);

  // Add the new service method to get item details
  rpc GetItemDetails(GetItemDetailsRequest) returns (GetItemDetailsResponse);

  // Add the new RPC method
  rpc UpdateItemDetails(ItemNotificationRequest2) returns (ItemNotificationResponse2);
}